feat: resolve server URLs before calling other listeners

sirdiego opened the pull request:
### Description

This PR moves the resolving of the URLs to before the server is started. This way one can use `configureServer` to get those information from the server.

```JavaScript
      configureServer(server) {
        server.httpServer.on('listening', () => {
          console.log(server.resolvedUrls)
        })
      },
```

I added a first test but am a little lost in how and where I could test the integrated functions. I'd love for this to get merged but would also appreciate help on how to write more tests here.

Thank you folks so much!

<!-- What is this PR solving? Write a clear description or reference the issues it solves (e.g. `fixes #123`). What other alternatives have you explored? Are there any parts you think require more attention from reviewers? -->

<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md.
- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Vite!
----------------------------------------------------------------------->

sirdiego committed:
feat: resolve server URLs before calling other listeners

sapphi-red commented on packages/vite/src/node/server/index.ts:
```diff
@@ -204,6 +205,7 @@ export interface ResolvedServerOptions
   > {
   fs: Required<FileSystemServeOptions>
   middlewareMode: NonNullable<ServerOptions['middlewareMode']>
+  hostname: Hostname
```
```suggestion
  /** @internal */
  hostname: Hostname
```
I'd like to mark this as a internal type.

sirdiego replied:
Makes sense, thanks for the feedback.

sirdiego committed:
test: add tests for the dev server

sirdiego commented:
> > I added a first test but am a little lost in how and where I could test the integrated functions. I'd love for this to get merged but would also appreciate help on how to write more tests here.
> 
> I think you can put it in `packages/vite/src/node/__tests__/dev.spec.ts` with something like
> 
> ```js
> test('check', () => {
>   expect.assertions(1)
>   const { promise, resolve } = promiseWithResolvers<void>()
>   const server = await createServer({
>     root: __dirname,
>     logLevel: 'error',
>     server: {
>       port: 9400, // make sure the port is unique
>       ws: false,
>     },
>     plugins: [/* add plugin that checks server.resolvedUrls here */] 
>   })
>   await server.listen()
>   await promise
> })
> ```
> 
> It needs to wait for the `configureServer` hook to be called so using `promiseWithResolvers` is needed. https://github.com/vitejs/vite/pull/19936/files#diff-0244b1dfcfa697a921ae7858a6beb979114a8c59b478ff360e72fdeb82b0b82cR226-R247

Thanks for the suggestion. That worked out very good. I added a new suite to the dev.spec.ts that fails without this patch and is green with this patch.

sapphi-red commented on packages/vite/src/node/__tests__/dev.spec.ts:
```diff
@@ -17,3 +20,62 @@ describe('resolveBuildEnvironmentOptions in dev', () => {
+    server?.close()
+  })
+
+  test.each([
+    [
+      {
+        port: 5013,
+      },
+      {
+        local: ['http://localhost:5013/'],
+        network: [],
+      },
+    ],
+    [
+      {
+        port: 5014,
+        host: '127.0.0.1',
+      },
+      {
+        local: ['http://127.0.0.1:5014/'],
+        network: [],
+      },
+    ],
+  ])('resolves the server URLs for %s', async (options, expectation) => {
```
```suggestion
  test('server.resolvedUrls exists at the time of `listening` event', async (options, expectation) => {
```
Since the purpose here is to test that `server.resolvedUrls` exists at the time of the `listening` event, I think testing just the first case is sufficient. Could you remove the second one?

sirdiego replied:
Thanks for the feedback. You are right, at this point it only matters to test if any host was resolved.

sirdiego commented on packages/vite/src/node/__tests__/dev.spec.ts:
```diff
@@ -17,3 +20,62 @@ describe('resolveBuildEnvironmentOptions in dev', () => {
+    await server.listen()
+    const urls = await promise
+
+    expect(urls).toMatchObject(expectation)
```
I did not know that, thanks for the feedback. Works with `toStrictEqual`.

sapphi-red commented on packages/vite/src/node/__tests__/dev.spec.ts:
```diff
@@ -17,3 +20,49 @@ describe('resolveBuildEnvironmentOptions in dev', () => {
+    expect.assertions(1)
+
+    const options = {
+      port: 5013,
```
```suggestion
      port: 5013, // make sure the port is unique
```

sapphi-red committed:
chore: add comment

sapphi-red reviewed (dismissed):
Thank you!

sirdiego commented:
Has the Windows test fail anything to do with these changes? Is there something I can do here?

sapphi-red commented:
It's probably a flaky fail. I've retriggered the CI 👍

sapphi-red committed:
chore: merge main
